Basic Implementation of the File System

JacobNickerson opened the pull request:
STRAIGHT TO MAIN

JacobNickerson committed:
added basic implementations of directories, files, and a file system API that should handle all calls from the shell, hopefully it works cause i didnt make the tests yet ;)

JacobNickerson committed:
added basic implementations of directories, files, and a file system API that should handle all calls from the shell, hopefully it works cause i didnt make the tests yet ;)

JacobNickerson committed:
Added incomplete unit testing to FileSystemTests

JacobNickerson committed:
Added incomplete unit testing to FileSystemTests

JacobNickerson committed:
Added more test cases for FileSystem

JacobNickerson committed:
Added more test cases for FileSystem

JacobNickerson committed:
Added new methods involving paths, fixed issues with findFSO

Added method overloading to the methods for creating files
and directories. Now, files and directories can be created by path.
This should be easier and more useful than having to find a reference
first. Also added a new method for changing the current directory,
eventually this should be moved to be a shell member, but for now
this works.

JacobNickerson committed:
Added new methods involving paths, fixed issues with findFSO

Added method overloading to the methods for creating files
and directories. Now, files and directories can be created by path.
This should be easier and more useful than having to find a reference
first. Also added a new method for changing the current directory,
eventually this should be moved to be a shell member, but for now
this works.

JacobNickerson committed:
Add new methods for creating directories and their unit tests

Added a new method override for createDirectory that accepts only
a path. This is more in line with the actual usage of 'mkdir' which
takes either just a directory name or a path with the directory name
at the end. Also added some minor methods like a getter for a
directory's children.

There are currently several unit tests for the new createDirectory
method, however, it most likely needs more unit tests as it is
possibly saucy.

JacobNickerson committed:
Add new methods for creating directories and their unit tests

Added a new method override for createDirectory that accepts only
a path. This is more in line with the actual usage of 'mkdir' which
takes either just a directory name or a path with the directory name
at the end. Also added some minor methods like a getter for a
directory's children.

There are currently several unit tests for the new createDirectory
method, however, it most likely needs more unit tests as it is
possibly saucy.

JacobNickerson committed:
FSAPI now a singleton, add some bash commands, FSAPI now in main

FSAPI's design pattern has been changed to be a Singleton, this was
actually done in the last commit, however I forgor to mention it > // <.
Now a reference to FSAPI can be obtained by calling the static method
`FileSystemAPI.getInstance()`.

FSAPI has been implemented into main. A few commands were also
implemented in order to test out the functionality of FSAPI. Namely
these commands are `cd`, `pwd`, and `mkdir`. Going forward, I will be
implementing bash commands in a different branch, as this one is only
concerned with the file system.

JacobNickerson committed:
FSAPI now a singleton, add some bash commands, FSAPI now in main

FSAPI's design pattern has been changed to be a Singleton, this was
actually done in the last commit, however I forgor to mention it > // <.
Now a reference to FSAPI can be obtained by calling the static method
`FileSystemAPI.getInstance()`.

FSAPI has been implemented into main. A few commands were also
implemented in order to test out the functionality of FSAPI. Namely
these commands are `cd`, `pwd`, and `mkdir`. Going forward, I will be
implementing bash commands in a different branch, as this one is only
concerned with the file system.

JacobNickerson committed:
This commit adds the ls command.

I couldn't help myself! I'm done for real this time.

cbloodsworth commented on src/main/java/jbash/Main.java:
```diff
@@ -14,11 +19,14 @@
         while (true) {
             // Prompt
-            System.out.print("[jbash]> ");
+            System.out.println("[jbash] " + (FSAPI.getCurrentDirectory().getPath()));
```
Good start for the prompt, but this should probably be one line (e.g. `[jbash] <pwd> $`).
I would probably want to abstract this out into its own module so down the line the user could modify it (think changing $PS1 in your `.bashrc`).

cbloodsworth commented on src/main/java/jbash/commands/CmdCd.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public int execute(List<String> args) {
+        if (args.size() > 1) { System.out.println("cd: too many arguments"); return 1; }
```
This reminded me -- we gotta do this #6

JacobNickerson replied:
Yeah honestly that would be better than this nonsense

cbloodsworth commented on src/main/java/jbash/filesystem/FileSystemObject.java:
```diff
@@ -1,11 +1,44 @@
 
-import java.nio.file.FileSystem;
-
 public abstract class FileSystemObject {
```
This should be package-private, since we don't want anyone else using this class: I'd change it to `abstract class FileSystemObject` (just remove the `public`.)

JacobNickerson replied:
where did that come from!

cbloodsworth commented on src/main/java/jbash/filesystem/FileSystemAPI.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    // TODO: Add support for quotes
+    public Optional<FileSystemObject> getFileSystemObject(String path) {
```
Javadoc comments for these methods would be really nice to see, especially if FileSystemAPI is front-facing:
https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#format

IntelliJ lets you just type `/**<enter>` above a method signature and it will fill out a boilerplate for you to fill in.

JacobNickerson replied:
im merging the branch ;) but i will include this afterwards. i'll make an issue

cbloodsworth commented on src/main/java/jbash/filesystem/FileSystemAPI.java:
```diff
@@ -0,0 +1,172 @@
+    public Optional<FileSystemObject> getFileSystemObject(String path) {
+        if (path.isEmpty()) { return Optional.empty(); }
+        boolean directorySearch = false;
+        if (path.endsWith("/")) { directorySearch = true; }  // Paths ending in / can only reference directories
```
```suggestion
        boolean directorySearch = path.endsWith("/");
```
May be more concise

cbloodsworth reviewed (approved):
Looks good to me. I added some comments for some stuff we'd probably want to change, but overall this is good.

JacobNickerson committed:
Adds changes suggested in code review

This commmit adds some of the suggestions given in the code review
for the file system pull request. However, FileSystemObject was not
made into a non-public class due to several tests failing. This
change will be explored further but will not be implemented in this
commit.

JacobNickerson committed:
Merge branch 'file-system-gaming' of github.com:JacobNickerson/jbash into file-system-gaming
